# Intro to Pandas (2) - Useful DataFrame Functions

In the last lesson we saw how to read data into a Pandas DataFrame from a .csv file on the internet or on your local machine and we also saw how to account for escape characters when reading a .csv file from a different folder in your local machine.

In this lesson we're going to look at some useful DataFrame functions and methods. Whilst we won't use these directly to plot data, utilising these will make your life a lot easier!

## Module Imports

In [1]:
#plotly.offline doesn't push your charts to the clouds
import plotly.offline as pyo
#allows us to create the Data and Figure objects
from plotly.graph_objs import *
#plotly.plotly pushes your charts to the cloud  
import plotly.plotly as py

#pandas is a data analysis library
import pandas as pd
from pandas import DataFrame

## Using DataFrames

Before we can learn about some useful DataFrame functions, we first need to create one. This time I'm going to assign the DataFrame to an object, rather than just display it.

We can see the DataFrame by typing the object name on its own line:

In [2]:
baseRateData = pd.read_csv("http://www.richard-muir.com/data/public/csv/BoEBaseRate.csv")

baseRateData

VALUE        DATE
0      11.50  02/01/1975
1      11.50  03/01/1975
2      11.50  06/01/1975
3      11.50  07/01/1975
4      11.50  08/01/1975
5      11.50  09/01/1975
6      11.50  10/01/1975
7      11.50  13/01/1975
8      11.50  14/01/1975
9      11.50  15/01/1975
10     11.50  16/01/1975
11     11.50  17/01/1975
12     11.25  20/01/1975
13     11.25  21/01/1975
14     11.25  22/01/1975
15     11.25  23/01/1975
16     11.25  24/01/1975
17     11.00  27/01/1975
18     11.00  28/01/1975
19     11.00  29/01/1975
20     11.00  30/01/1975
21     11.00  31/01/1975
22     11.00  03/02/1975
23     11.00  04/02/1975
24     11.00  05/02/1975
25     11.00  06/02/1975
26     11.00  07/02/1975
27     10.75  10/02/1975
28     10.75  11/02/1975
29     10.75  12/02/1975
...      ...         ...
10456   0.50  12/05/2016
10457   0.50  13/05/2016
10458   0.50  16/05/2016
10459   0.50  17/05/2016
10460   0.50  18/05/2016
10461   0.50  19/05/2016
10462   0.50  20/05/2016
10463   0.50  23/05/2016
10464   0.50  24/05/2016
10465   0.50  25/05/2016
10466   0.50  26/05/2016
10467   0.50  27/05/2016
10468   0.50  31/05/2016
10469   0.50  01/06/2016
10470   0.50  02/06/2016
10471   0.50  03/06/2016
10472   0.50  06/06/2016
10473   0.50  07/06/2016
10474   0.50  08/06/2016
10475   0.50  09/06/2016
10476   0.50  10/06/2016
10477   0.50  13/06/2016
10478   0.50  14/06/2016
10479   0.50  15/06/2016
10480   0.50  16/06/2016
10481   0.50  17/06/2016
10482   0.50  20/06/2016
10483   0.50  21/06/2016
10484   0.50  22/06/2016
10485   0.50  23/06/2016

[10486 rows x 2 columns]

This DataFrame is really long, and it's a pain to keep scrolling up to see the column names. We can use <code>df.head(x)</code> to see the top x rows (df is the conventional abbreviation for DataFrame, which I'll use throughout the course). 

The default number of rows is 5. Note that this doesn't change the DataFrame, merely how it is displayed on the screen.

In [3]:
baseRateData.head(10)

VALUE        DATE
0   11.5  02/01/1975
1   11.5  03/01/1975
2   11.5  06/01/1975
3   11.5  07/01/1975
4   11.5  08/01/1975
5   11.5  09/01/1975
6   11.5  10/01/1975
7   11.5  13/01/1975
8   11.5  14/01/1975
9   11.5  15/01/1975

In [4]:
baseRateData.head()

VALUE        DATE
0   11.5  02/01/1975
1   11.5  03/01/1975
2   11.5  06/01/1975
3   11.5  07/01/1975
4   11.5  08/01/1975

We can also see the bottom x rows using <code>df.tail(x)</code>. Once again, the default is 5.

In [5]:
baseRateData.tail(5)

VALUE        DATE
10481    0.5  17/06/2016
10482    0.5  20/06/2016
10483    0.5  21/06/2016
10484    0.5  22/06/2016
10485    0.5  23/06/2016

That's much more manageable!

Before we start looking at more useful functions, let's take a quick tour of a DataFrame.

You can see that this DataFrame has 10,486 rows (remember, Python is zero-indexed!) and 3 columns. The left-most column (in bold) is called the index. The index of the DataFrame doesn't necessarily hold any data; instead it allows us to access the data by its location.

We can also see the column names of the DataFrame. Used in conjunction with the index, we can reference an exact cell within the DataFrame.

We can get some more information about the DataFrame by using the <code>df.info()</code> method.

In [6]:
baseRateData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10486 entries, 0 to 10485
Data columns (total 2 columns):
VALUE    10486 non-null float64
DATE     10486 non-null object
dtypes: float64(1), object(1)
memory usage: 163.9+ KB


This tells us that we are inspecting a DataFrame and that the DataFrame has 10,486 rows (indexed from 0-10,485), and 2 data columns (the index doesn't hold data). It also gives us some information about the type of data stored in each column, as well as the approximate amount of RAM used to hold it.

We can get some statistical information on numeric columns by using the <code>df.describe()</code> method:

In [7]:
baseRateData.describe()

VALUE
count  10486.000000
mean       7.135085
std        4.515003
min        0.500000
25%        4.500000
50%        6.125000
75%       10.562500
max       17.000000

This has given us some summary statistics for the <code>VALUE</code> column, for example we can see that the average base rate for the time period is 7.14%. This is useful to allow you to get an idea of the shape of your data - if it does not look as you expect after plotting it, something may have gone wrong along the way!

Finally, we can get the column names in a DataFrame and store it in a list. This is helpful when we want to plot different data columns on the same chart - we can loop through the column names.

In [8]:
baseRateData.columns.tolist()

['VALUE', 'DATE']

### What have we learnt this lesson?

In this lesson we've seen how to restrict the amount of rows that are shown when we view the DataFrame using <code>df.head()</code> and <code>df.tail()</code>. We've also learnt how to describe the DataFrame object, as well as how to obtain some summary statistics.

In the next lesson we'll see how to select and access data from a DataFrame.

If you have any questions, please ask in the comments section or email <a href="mailto:me@richard-muir.com">me@richard-muir.com</a>